Initial support for SBOM generation and CycloneDX

aloubyansky opened the pull request.

aloubyansky committed:
Initial support for SBOM generation and CycloneDX

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

Co-authored-by: Rostislav Svoboda <rsvoboda@redhat.com>

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+* `quarkusDevRuntimeClasspathConfiguration` - Quarkus application dev mode runtime dependencies;
+* `quarkusDevRuntimeClasspathConfigurationDeployment` - Quarkus application dev mode runtime and build time dependencies.
+
+Given that the plugin is not aware of how Quarkus uses these dependencies, it will not be able to set `quarkus:component:scope` property for components. On the other hand, the requested configuration name can be used indicate which scope to target.
```
```suggestion
Given that the plugin is not aware of how Quarkus uses these dependencies, it will not be able to set the `quarkus:component:scope` property for components. On the other hand, the requested configuration name can be used indicate which scope to target.
```

gsmet replied:
Yes because here it's definite, it's this particular property you want to set.

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+
+This chapter describes SBOMs that manifest outcomes of Quarkus builds that are final application distributions.
+
+During an application build and package assembly process Quarkus captures certain details about the produced distribution and then allows an SBOM generator to consume and record that information in an SBOM format.
```
```suggestion
During an application build and package assembly process, Quarkus captures certain details about the produced distribution and then allows an SBOM generator to consume and record that information in an SBOM format.
```

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+
+==== Runtime Components
+
+Every file in the resulting Fast JAR distribution will appear in the SBOM with `quarkus:component:scope` property set to `runtime` and `evidence.occurences.location` field pointing to the location of the component in the application distribution directory, for example
```
```suggestion
Every file in the resulting Fast JAR distribution will appear in the SBOM with the `quarkus:component:scope` property set to `runtime` and `evidence.occurences.location` field pointing to the location of the component in the application distribution directory, for example
```

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+      }
+----
+
+NOTE: `evidence.occurrences.location` was introduced in CycloneDX schema version 1.5, for older versions the location will be indicated using `quarkus:component:location` property.
```
```suggestion
NOTE: `evidence.occurrences.location` was introduced in CycloneDX schema version 1.5, for older versions the location will be indicated using the `quarkus:component:location` property.
```

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+
+Pedigree is a way to provide information that certain patches, or changes in general, have been applied to a certain component.
+
+In certain cases Quarkus may copy modified versions of dependency artifacts to an application distribution. Manipulating original content of a component will change its hash sums which may get highlighted as suspicious by the tools comparing original component hash sums to those found in the distribution.
```
I wonder if `checksums` would be better than `hash sums`? I.e. more commonly used.

gsmet replied:
OK, I suppose it's fine then.

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+
+In certain cases Quarkus may copy modified versions of dependency artifacts to an application distribution. Manipulating original content of a component will change its hash sums which may get highlighted as suspicious by the tools comparing original component hash sums to those found in the distribution.
+
+When Quarkus applies modifications to artifacts resolved from Maven repositories it can manifest these changes as pedigree notes in the generated SBOM.
```
```suggestion
When Quarkus applies modifications to artifacts resolved from Maven repositories, it can manifest these changes as pedigree notes in the generated SBOM.
```

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+
+==== Build time dependencies
+
+Build time dependencies will be recorded with `quarkus:component:scope` property set to `development`:
```
```suggestion
Build time dependencies will be recorded with the `quarkus:component:scope` property set to `development`:
```

gsmet commented on core/deployment/banned-signatures.txt:
```diff
@@ -1,6 +1,6 @@
-@defaultMessage Don't use Maven classes. They won't be available when using Gradle.
-org.apache.maven.**
-org.codehaus.plexus.**
+#@defaultMessage Don't use Maven classes. They won't be available when using Gradle.
+#org.apache.maven.**
+#org.codehaus.plexus.**
```
Is it actually needed? I don't see any of these packages added to `core/deployment`? But I don't remember if forbidden-apis also scans the classpath.

gsmet reviewed (commented):
Interesting work! It's nice to see this coming to fruition.

I added some comments/questions/suggestions here and there.

I have one more question though: does it add some significant work to standard dev mode build?

gsmet commented on extensions/cyclonedx/deployment/pom.xml:
```diff
@@ -0,0 +1,53 @@
+
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <annotationProcessorPaths>
+                        <path>
+                            <groupId>io.quarkus</groupId>
+                            <artifactId>quarkus-extension-processor</artifactId>
+                            <version>${project.version}</version>
+                        </path>
+                    </annotationProcessorPaths>
+                </configuration>
+            </plugin>
```
```suggestion
            <plugin>
                <artifactId>maven-compiler-plugin</artifactId>
                <executions>
                    <execution>
                        <id>default-compile</id>
                        <configuration>
                            <annotationProcessorPaths>
                                <path>
                                    <groupId>io.quarkus</groupId>
                                    <artifactId>quarkus-extension-processor</artifactId>
                                    <version>${quarkus.version}</version>
                                </path>
                            </annotationProcessorPaths>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

gsmet commented on extensions/cyclonedx/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,7 @@
+  keywords:
+  - "cyclonedx"
+  - "cdx"
+  status: "preview"
```
Given you added a guide, you could add it there too.

Example here: https://github.com/quarkusio/quarkus/blob/main/extensions/hibernate-validator/runtime/src/main/resources/META-INF/quarkus-extension.yaml#L10

gsmet reviewed (commented):
Added two more minor comments.

aloubyansky commented on independent-projects/bootstrap/app-model/src/main/java/io/quarkus/sbom/ApplicationManifestConfig.java:
```diff
@@ -0,0 +1,194 @@
+public class ApplicationManifestConfig {
+
+    public static Builder builder() {
+        return new ApplicationManifestConfig().new Builder();
```
It's an implementation detail anyway

aloubyansky commented on extensions/cyclonedx/deployment/src/main/java/io/quarkus/cyclonedx/CycloneDxConfig.java:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @return CycloneDX specification version
+     */
+    @WithDefault("1.6")
```
I changed it to default to the latest the library supports.

gsmet commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,183 @@
+
+`XML` format can be requested by setting `format` goal parameter (or `quarkus.dependency.sbom.format` property) to `xml`.
+
+Each component in the generated SBOM will include `quarkus:component:scope` property that will indicate whether this component is used at runtime or only development/build time.
```
I think both are correct in this case.

gsmet replied:
Indefinite or definite will work fine here, I let you choose the one you prefer :).

aloubyansky commented on build-parent/pom.xml:
```diff
@@ -40,6 +40,7 @@
         <!-- Dev tools -->
         <freemarker.version>2.3.33</freemarker.version>
         <commonmark.version>0.22.0</commonmark.version>
+        <cyclonedx.version>9.0.5</cyclonedx.version>
```
Added it to the quarkus-bom

aloubyansky commented:
@gsmet I resolved most of your suggestions, could you check it once again? I didn't go with the record, wasn't sure about that.

aloubyansky commented:
I'll look into adding proper tests next

gsmet commented:
> I didn't go with the record, wasn't sure about that.

I can perfectly understand that :).

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+
+== Why Quarkus-specific tooling?
+
+While Quarkus integrates with build tools such as https://maven.apache.org/[Maven] and https://gradle.org/[Gradle], it could itself be categorized as a build tool with its own component and dependency model, build steps and build outcomes. One of the essential component types of a Quarkus application is a Quarkus extension, which consists of a runtime and a build time artifacts, and their depende ...
```
```suggestion
While Quarkus integrates with build tools such as https://maven.apache.org/[Maven] and https://gradle.org/[Gradle], it could itself be categorized as a build tool with its own component and dependency model, build steps, and build outcomes. One of the essential component types of a Quarkus application is a Quarkus extension, which consists of a runtime and a build time artifacts, and their dependencies.
```

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+
+To properly resolve Quarkus extension and other application dependencies Quarkus uses its own dependency resolver, which is implemented on top of the dependency resolver provided by the underlying build tool: Maven or Gradle.
+
+As a consequence, in case of Maven, for example, the results of `dependency:tree` will not include all the dependencies Quarkus will use to build an application. A similar issue will affect other dependency analysis tools that assume a project adheres to the standard Maven dependency model: they will not be able to capture the effective Quarkus application dependency graph. That, unfortunately, i ...
```
```suggestion
As a consequence, in case of Maven, for example, the results of `dependency:tree` will not include all the dependencies Quarkus will use to build an application. A similar issue will affect other dependency analysis tools that assume a project adheres to the standard Maven dependency model: they will not be able to capture the effective Quarkus application dependency graph. Unfortunately, that includes the implementation of the https://github.com/CycloneDX/cyclonedx-maven-plugin[CycloneDX Maven plugin].
```

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+
+As a consequence, in case of Maven, for example, the results of `dependency:tree` will not include all the dependencies Quarkus will use to build an application. A similar issue will affect other dependency analysis tools that assume a project adheres to the standard Maven dependency model: they will not be able to capture the effective Quarkus application dependency graph. That, unfortunately, i ...
+
+Besides the dependencies, that are an input to a build process, there is also an outcome of the build that is the final distribution of an application. Users of an application may request an SBOM manifesting not only the dependencies (the input to a build) but also the final distribution (the outcome of the build) before they agree to deploy the application. Quarkus allows application developers  ...
```
```suggestion
Besides the dependencies, that are an input to a build process, there is also an outcome of the build that is the final distribution of an application. Users of an application may request an SBOM manifesting not only the dependencies (the input to a build) but also the final distribution (the outcome of the build) before they agree to deploy the application. Quarkus allows application developers to choose various packaging types for their applications, some of which are Quarkus-specific. Providing certain Quarkus-specific details about components included in a distribution may help better evaluate the impact of potential security-related issues.
```

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+
+=== Maven Dependency SBOMs
+
+For Quarkus Maven projects dependency SBOMs can be generated with the `quarkus:dependency-sbom` goal. The outcome of the goal will be saved in a `target/<artifactId>-<version>-dependency-cyclonedx.json` file (which can be changed by setting the `outputFile` goal parameter or the `quarkus.dependency.sbom.output-file` property). The complete Quarkus build and runtime dependency graphs will be recod ...
```
```suggestion
For Quarkus Maven projects dependency SBOMs can be generated with the `quarkus:dependency-sbom` goal. The outcome of the goal will be saved in a `target/<artifactId>-<version>-dependency-cyclonedx.json` file (which can be changed by setting the `outputFile` goal parameter or the `quarkus.dependency.sbom.output-file` property). The complete Quarkus build and runtime dependency graphs will be recorded in the https://cyclonedx.org/[CycloneDX] `JSON` format.
```

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+        }
+----
+
+By default, `quarkus:dependency-sbom` captures the dependencies of a production build. Quarkus supports three application bootstrap modes: normal (production), test and dev. In all three modes an application may have different dependency graphs. The `mode` parameter can be used to indicate which dependency graph should be recorded. If the `mode` is set to `test` or `dev`, the target file name wil ...
```
```suggestion
By default, `quarkus:dependency-sbom` captures the dependencies of a production build. Quarkus supports three application bootstrap modes: normal (production), test, and dev. In all three modes, an application may have different dependency graphs. The `mode` parameter can be used to indicate which dependency graph should be recorded. If the `mode` is set to `test` or `dev`, the target file name will become `target/<artifactId>-<version>-<mode>-dependency-cyclonedx.json`.
```

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+
+==== Runtime Components
+
+Every file in the resulting Fast JAR distribution will appear in the SBOM with the `quarkus:component:scope` property set to `runtime` and `evidence.occurences.location` field pointing to the location of the component in the application distribution directory, for example
```
Shouldn't be there `occurrences` - double `r`?

It's in multiple places.

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+
+Pedigree is a way to provide information that certain patches, or changes in general, have been applied to a certain component.
+
+In certain cases Quarkus may copy modified versions of dependency artifacts to an application distribution. Manipulating original content of a component will change its hash sums which may get highlighted as suspicious by the tools comparing original component hash sums to those found in the distribution.
```
```suggestion
In certain cases, Quarkus may copy modified versions of dependency artifacts to an application distribution. Manipulating the original content of a component will change its hash sums which may get highlighted as suspicious by the tools comparing original component hash sums to those found in the distribution.
```

rsvoboda commented on docs/src/main/asciidoc/cyclonedx.adoc:
```diff
@@ -0,0 +1,192 @@
+|Name |Value range |Description
+
+|`quarkus:component:scope` |`runtime` or `development` |Indicates whether a component is a runtime or a build/development time dependency of an application.
+|`quarkus:component:location` |String representing a file system path using `/` as a path element |Used in SBOMs with schema versions 1.4 or older. Starting from schema 1.5, `evidence.occurrences.location` is used instead. This property is used only if a component in found in the distribution. The value is a relative path to a file pointing to the location of a component in a distribution using ` ...
```
```suggestion
|`quarkus:component:location` |String representing a file system path using `/` as a path element |Used in SBOMs with schema versions 1.4 or older. Starting from schema 1.5, `evidence.occurrences.location` is used instead. This property is used only if a component is found in the distribution. The value is a relative path to a file pointing to the location of a component in a distribution using `/` as a path element separator.
```

rsvoboda commented:
@aloubyansky I did look into docs for now, I will check the functionality later today.

rsvoboda reviewed (changes requested):
I did some experiments, some things need adjustments, some things are discussion topics.

1) Silent `mvn quarkus:dependency-sbom`
`mvn quarkus:dependency-sbom` could report in the log that `target/code-with-quarkus-1.0.0-SNAPSHOT-dependency-cyclonedx.json` was generated

2) Strange `Allowed values are {code json}` in help:describe
```
mvn help:describe -Dcmd=quarkus:dependency-sbom -Ddetail

    format (Default: json)
      User property: quarkus.dependency.sbom.format
      CycloneDX BOM format. Allowed values are {code json} and xml. The default
      is json.
```

3) Mode has user property mode, not prefixed with quarkus.dependency.sbom.
```
mvn help:describe -Dcmd=quarkus:dependency-sbom -Ddetail

   mode (Default: prod)
      User property: mode
```
The other sbom related properties are prefixed with quarkus.dependency.sbom

4) quarkus-cyclonedx dependency - generated file name
```
mvn clean package -Dnative && ll target | grep cyclo
-rw-r--r--@ 1 rsvoboda  staff   659K Aug 13 10:33 code-with-quarkus-1.0.0-SNAPSHOT-runner-cyclonedx.json
-rw-r--r--@ 1 rsvoboda  staff   687K Aug 13 10:33 quarkus-run-cyclonedx.json

```
In JVM mode the file name will always start with `quarkus-run`, I would personally prefer name based on the artifactId+version as in case of native or `mvn quarkus:dependency-sbom`

5) SBOM for JVM vs Native
`cyclonedx.adoc` could have a note that the SBOM for Native and JVM mode will be different because of all the details that are captured within the file.
I initially thought they would be the same, but after deeper checks of the file content I realized they will always differ - e.g. components, evidence details.

6) cyclonedx.adoc - Fast JAR
I would put it as first, why is Uber JAR and Native before it?
Add a note that this is the default packaging for Quarkus to make it more obvious for readers.

7) cyclonedx.adoc - Uber JAR
I'm confused by "Since an Uber JAR is a Maven artifact itself, SBOMs generated for Uber JARs will also be automatically attached as project artifacts by the Quarkus Maven plugin. This, however, can be disabled by setting attachSboms parameter of the quarkus:build goal to false."
`automatically attached` ... I realized that `mvn clean install` puts sboms into my `~/.m2`, but after I played with uber jar, it wasn't clear to me what exactly is meant by that line. I thought it's attached into the uber jar under META-INF ...
`setting attachSboms parameter of the quarkus:build goal` ... majority of people do `mvn clean install` or `mvn clean package`, how does attachSboms parameter translate to this flow, where do I specify that property?

8) quarkus.cyclonedx.format=xmlds
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project code-with-quarkus: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.cyclonedx.deployment.CdxSbomBuildStep#generate threw an exception: java.lang.RuntimeException: Unsupported SBOM artifact type xmlds, supported types are json and xml
```
I think this is OK, just adding for the information that some negative use-cases were tried.

9) Fast JAR - Build time dependencies - option to skip them
Can be the build time dependencies dropped from the SBOM file so just runtime details are covered?

10) https://github.com/CycloneDX/cyclonedx-cli experiments
Unhandled exception: System.ArgumentException: Unsupported specification version: 1.6

Is there a reason to be on 1.6 version of the specification?

I have manually adjusted the file to be 1.5 and I could run commands `cyclonedx diff target/quarkus-run-cyclonedx.xml target/code-with-quarkus-1.0.0-SNAPSHOT-runner-cyclonedx.xml --component-versions` and `cyclonedx analyze target/quarkus-run-cyclonedx.xml` without any problem

11) NO TESTS in the PR
There are no tests ...

12) components - hashes in SBOM file
There are 8 variants of hashes for every component. What's the motivation to cover so many algorithms ?

13) Usage with quarkus extensions
I was able to get just the dependency SBOM through quarkus maven plugin. Usage of `io.quarkus:quarkus-cyclonedx` artifact is not triggering any report during the build.
Documentation should highlight that the usa-case is for Quarkus applications, not for the extensions.
Do you expect/plan to be able to generate SBOMs for extensions?

14) Multi-module project support
I played with https://github.com/quarkus-qe/beefy-scenarios and added `io.quarkus:quarkus-cyclonedx` artifact. There was `quarkus-run-cyclonedx.json` files generated in target directory of every module.

I would appreciate:
 - some aggregated SBOM summary for all the deps used across the modules
 - The same `quarkus-run-cyclonedx.json` naming across all the modules is not ideal for me, I would appreciate naming based on the artifactId and maybe version to avoid file name collision when for example extracting the files from the project before uploading into dependencytrack

aloubyansky commented:
> I did some experiments, some things need adjustments, some things are discussion topics.
> 
> 1. Silent `mvn quarkus:dependency-sbom`
>    `mvn quarkus:dependency-sbom` could report in the log that `target/code-with-quarkus-1.0.0-SNAPSHOT-dependency-cyclonedx.json` was generated

Fixed by logging
```
[INFO] --- quarkus:999-SNAPSHOT:dependency-sbom (default-cli) @ test-app ---
[INFO] The SBOM has been saved in {absolute-path}/test-app/target/test-app-1.0.0-SNAPSHOT-dependency-cyclonedx.json
```

> 2. Strange `Allowed values are {code json}` in help:describe
> 
> ```
> mvn help:describe -Dcmd=quarkus:dependency-sbom -Ddetail
> 
>     format (Default: json)
>       User property: quarkus.dependency.sbom.format
>       CycloneDX BOM format. Allowed values are {code json} and xml. The default
>       is json.
> ```

Fixed, missed a symbol. Thanks.

> 3. Mode has user property mode, not prefixed with quarkus.dependency.sbom.
> 
> ```
> mvn help:describe -Dcmd=quarkus:dependency-sbom -Ddetail
> 
>    mode (Default: prod)
>       User property: mode
> ```
> 
> The other sbom related properties are prefixed with quarkus.dependency.sbom

Yeah, I copied it from `quarkus:dependency-tree`. But fixed it in this goal to be consistent with the rest of the properties. Thanks.

> 4. quarkus-cyclonedx dependency - generated file name
> 
> ```
> mvn clean package -Dnative && ll target | grep cyclo
> -rw-r--r--@ 1 rsvoboda  staff   659K Aug 13 10:33 code-with-quarkus-1.0.0-SNAPSHOT-runner-cyclonedx.json
> -rw-r--r--@ 1 rsvoboda  staff   687K Aug 13 10:33 quarkus-run-cyclonedx.json
> ```
> 
> In JVM mode the file name will always start with `quarkus-run`, I would personally prefer name based on the artifactId+version as in case of native or `mvn quarkus:dependency-sbom`

There are a few points here to make.

1) there could be multiple packages produced by a single Quarkus build. For example, if you simply build with `-Dnative`, it will enable the native build but will not disable the JAR build. So, in fact, the outcome of that will be two executables and two SBOMs. The issue is that, not every packaging format
```
drwxr-xr-x. 5 aloubyansky aloubyansky     4096 Aug 13 12:50 quarkus-app
-rw-r--r--. 1 aloubyansky aloubyansky   725074 Aug 13 12:52 quarkus-run-cyclonedx.json
-rwxr-xr-x. 1 aloubyansky aloubyansky 52416880 Aug 13 12:52 test-app-1.0.0-SNAPSHOT-runner
-rw-r--r--. 1 aloubyansky aloubyansky   695769 Aug 13 12:52 test-app-1.0.0-SNAPSHOT-runner-cyclonedx.json
```
So an SBOM file name is derived from the name of the executable it manifests.

2) Not every built executable is a Maven artifact. ATM, only the Uber JAR is by default attached as a Maven artifact to the project and so its SBOM.

At least these kinds of thoughts were what led to the current implementation. Let me know in case you have better ideas how to deal with this. Thanks.

> 5. SBOM for JVM vs Native
>    `cyclonedx.adoc` could have a note that the SBOM for Native and JVM mode will be different because of all the details that are captured within the file.
>    I initially thought they would be the same, but after deeper checks of the file content I realized they will always differ - e.g. components, evidence details.

In terms of components, Uber JAR should be identical to Native actually. A directory-based distribution will surely be different to a single file that bundles everything. In fact, I'd expect every distribution to be somewhat different. There are actually others that I haven't covered yet.
I added a few words about `evidence.occurrences.location` to Uber JAR and Native. 

> 6. cyclonedx.adoc - Fast JAR
>    I would put it as first, why is Uber JAR and Native before it?
>    Add a note that this is the default packaging for Quarkus to make it more obvious for readers.

Sure, moved it up.

> 7. cyclonedx.adoc - Uber JAR
>    I'm confused by "Since an Uber JAR is a Maven artifact itself, SBOMs generated for Uber JARs will also be automatically attached as project artifacts by the Quarkus Maven plugin. This, however, can be disabled by setting attachSboms parameter of the quarkus:build goal to false."
>    `automatically attached` ... I realized that `mvn clean install` puts sboms into my `~/.m2`, but after I played with uber jar, it wasn't clear to me what exactly is meant by that line. I thought it's attached into the uber jar under META-INF ...
>    `setting attachSboms parameter of the quarkus:build goal` ... majority of people do `mvn clean install` or `mvn clean package`, how does attachSboms parameter translate to this flow, where do I specify that property?

"Attaching an artifact to a project" is a Maven terminology, at least that's how the relevant API is named when extra artifacts need to be published along with the default artifacts.

As the sentence says `attachSboms` is a parameter of the `quarkus:build` goal and so it has to be set in the plugin configuration or with the corresponding property.

Since it's about publishing artifacts to Maven repositories, `package` isn't really affected by this but install is.

```
$ mvn clean install -DskipTests -Dquarkus.package.jar.type=uber-jar -Dquarkus.cyclonedx.format=all
...
[INFO] --- install:3.1.1:install (default-install) @ test-app ---
[INFO] Installing /home/aloubyansky/playground/test-app/pom.xml to /home/aloubyansky/.m2/repository/org/acme/test-app/1.0.0-SNAPSHOT/test-app-1.0.0-SNAPSHOT.pom
[INFO] Installing /home/aloubyansky/playground/test-app/target/test-app-1.0.0-SNAPSHOT.jar.original to /home/aloubyansky/.m2/repository/org/acme/test-app/1.0.0-SNAPSHOT/test-app-1.0.0-SNAPSHOT.jar
[INFO] Installing /home/aloubyansky/playground/test-app/target/test-app-1.0.0-SNAPSHOT-runner.jar to /home/aloubyansky/.m2/repository/org/acme/test-app/1.0.0-SNAPSHOT/test-app-1.0.0-SNAPSHOT-runner.jar
[INFO] Installing /home/aloubyansky/playground/test-app/target/test-app-1.0.0-SNAPSHOT-runner-cyclonedx.json to /home/aloubyansky/.m2/repository/org/acme/test-app/1.0.0-SNAPSHOT/test-app-1.0.0-SNAPSHOT-cyclonedx.json
[INFO] Installing /home/aloubyansky/playground/test-app/target/test-app-1.0.0-SNAPSHOT-runner-cyclonedx.xml to /home/aloubyansky/.m2/repository/org/acme/test-app/1.0.0-SNAPSHOT/test-app-1.0.0-SNAPSHOT-cyclonedx.xml
```

I re-phrased the text a bit. Hope that helps.

> 8. quarkus.cyclonedx.format=xmlds
> 
> ```
> [ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project code-with-quarkus: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
> [ERROR] 	[error]: Build step io.quarkus.cyclonedx.deployment.CdxSbomBuildStep#generate threw an exception: java.lang.RuntimeException: Unsupported SBOM artifact type xmlds, supported types are json and xml
> ```
> 
> I think this is OK, just adding for the information that some negative use-cases were tried.
> 
> 9. Fast JAR - Build time dependencies - option to skip them
>    Can be the build time dependencies dropped from the SBOM file so just runtime details are covered?

Sure. I'd do that in a follow up PR though.

> 10. https://github.com/CycloneDX/cyclonedx-cli experiments
>     Unhandled exception: System.ArgumentException: Unsupported specification version: 1.6
> 
> Is there a reason to be on 1.6 version of the specification?

No besides it being the latest version of the spec. It can be specified with `quarkus.cyclonedx.schema-version`.

> I have manually adjusted the file to be 1.5 and I could run commands `cyclonedx diff target/quarkus-run-cyclonedx.xml target/code-with-quarkus-1.0.0-SNAPSHOT-runner-cyclonedx.xml --component-versions` and `cyclonedx analyze target/quarkus-run-cyclonedx.xml` without any problem
> 
> 11. NO TESTS in the PR
>     There are no tests ...

Yes, as I mentioned, I'll add them in a follow up PR, just to put a foot into the upcoming CR1 :)

> 12. components - hashes in SBOM file
>     There are 8 variants of hashes for every component. What's the motivation to cover so many algorithms ?

Yes, I got the same remark from our colleagues from OWASP, specially about SHA3 algorithms, but I use the hash utility from the cyclonedx-core-java library and that's what it produces, the same is used in the cyclonedx-maven-plugin.

> 13. Usage with quarkus extensions
>     I was able to get just the dependency SBOM through quarkus maven plugin. Usage of `io.quarkus:quarkus-cyclonedx` artifact is not triggering any report during the build.
>     Documentation should highlight that the usa-case is for Quarkus applications, not for the extensions.
>     Do you expect/plan to be able to generate SBOMs for extensions?

Given that every Quarkus extension is meant to be used for application development, I'd expect it to be a given.
No, I haven't yet thought about extension development in this regard.

> 14. Multi-module project support
>     I played with https://github.com/quarkus-qe/beefy-scenarios and added `io.quarkus:quarkus-cyclonedx` artifact. There was `quarkus-run-cyclonedx.json` files generated in target directory of every module.
> I would appreciate:
> 
> * some aggregated SBOM summary for all the deps used across the modules

That's out of scope of this effort. At this point we are covering Quarkus applications only.

> * The same `quarkus-run-cyclonedx.json` naming across all the modules is not ideal for me, I would appreciate naming based on the artifactId and maybe version to avoid file name collision when for example extracting the files from the project before uploading into dependencytrack

When it comes to naming conflicts, the same could be said about `quarkus-run` as a name for a distribution. This output isn't expected to be unique across modules.
I am not sure what a reasonable naming convention would be keeping in mind the fact that the same module can produce multiple application builds.

rsvoboda commented:
> quarkus-run-cyclonedx.json naming

OK, let's keep it as is, we can adjust later if needed.

> No besides it being the latest version of the spec. It can be specified with quarkus.cyclonedx.schema-version.

Ok, I didn't find it mentioned in the documentation. This could be there as people can be using different tools.

One tiny concern I have is the inconsistency of the schema-version property names, but not sure what to do about it.
 - quarkus:dependency-sbom needs `quarkus.dependency.sbom.schema-version=1.5`
 - quarkus-cyclonedx needs `quarkus.cyclonedx.schema-version=1.5`

Maybe just using `quarkus.sbom` prefix? We may end up supporting different format, not just cyclonedx ... so we would still need `quarkus.cyclonedx.` configs ... probably not worth reworking what's available in the PR

rsvoboda reviewed (approved):
Let's go with this. We will tune it as we go.
